ocaml/xapi/rpm_gpg_key.ml

```diff
@@ -0,0 +1,246 @@
+(*
+ * Copyright (C) 2022 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "rpm_gpg_key" end)
+
+open D
+
+let assert_name_is_valid ~name =
+  let safe_char = function
+    | 'A' .. 'Z' | 'a' .. 'z' | '0' .. '9' | '.' | '_' | '-' ->
+        true
+    | _ ->
+        false
+  in
+  let len = String.length name in
+  let starts_with_dot = Astring.String.is_prefix ~affix:"." name in
+  let is_valid = String.for_all safe_char name && not starts_with_dot in
+  match (is_valid, len <= 0, len > 64) with
+  | false, _, _ ->
+      raise Api_errors.(Server_error (rpm_gpg_key_name_is_invalid, [name]))
+  | true, true, _ ->
+      raise Api_errors.(Server_error (rpm_gpg_key_name_is_empty, []))
+  | true, false, true ->
+      raise Api_errors.(Server_error (rpm_gpg_key_name_is_too_long, [name]))
+  | true, false, false ->
+      ()
+
+module PubKeyMetaData = struct
+  type t = {created: float; fingerprint: string}
+end
+
+let parse_pubkey_metadata md_str_in_colons =
+  let open Angstrom in
+  let s = char ':' in
+  let p = take_till (fun x -> x = ':') in
+  let colons = sep_by s p >>= fun ss -> return ss in
+  let consume = Consume.Prefix in
+  let r = (parse_string ~consume colons) md_str_in_colons in
+  match r with
+  | Ok l ->
```

Comment:
Would it not be easier to match to use Angstrom combinators like `count` to consume the expected number of `:` and let the parser fails rather than doing this manually here? I would also try to let the parser return a list and match the structure of this list here and avoid using something like `List.nth`

Reply:
Thanks. Yeah, using `count` looks better:
```
  let single_field =
    take_while (fun x -> x <> ':') >>= fun v ->
    char ':' >>= fun _ ->
    return v
  in
  let pickup =
    single_field >>= fun _ ->
    count 4 single_field >>= fun _ ->
    single_field >>= fun created ->
    count 13 single_field >>= fun _ ->
    single_field >>= fun fingerprint  ->
    return (created, fingerprint)
  in
```